field/binary.go

```diff
@@ -120,34 +117,82 @@ func (f *Binary) SetData(data interface{}) error {
 }
 
 func (f *Binary) Unmarshal(v interface{}) error {
-	if v == nil {
-		return nil
+	switch val := v.(type) {
+	case reflect.Value:
+		switch val.Kind() { //nolint:exhaustive
+		case reflect.String:
+			if !val.CanSet() {
+				return fmt.Errorf("reflect.Value of the data can not be change")
+			}
+
+			str := hex.EncodeToString(f.value)
+			val.SetString(str)
+		case reflect.Slice:
+			if !val.CanSet() {
+				return fmt.Errorf("reflect.Value of the data can not be change")
+			}
+
+			val.SetBytes(f.value)
+		default:
+			return fmt.Errorf("data does not match required reflect.Value type")
+		}
+	case *string:
+		str := hex.EncodeToString(f.value)
+		*val = str
+	case *[]byte:
+		*val = f.value
+	case *Binary:
+		val.value = f.value
+	default:
+		return fmt.Errorf("data does not match required *Binary or (*string, *[]byte) type")
 	}
 
-	bin, ok := v.(*Binary)
-	if !ok {
-		return errors.New("data does not match required *Binary type")
-	}
-
-	bin.value = f.value
-
 	return nil
 }
 
 func (f *Binary) Marshal(v interface{}) error {
-	if v == nil {
-		return nil
-	}
-
-	bin, ok := v.(*Binary)
-	if !ok {
-		return errors.New("data does not match required *Binary type")
+	switch v := v.(type) {
+	case *Binary:
+		if v == nil {
+			return nil
+		}
+		f.value = v.value
+	case string:
+		if v == "" {
+			f.value = nil
+			return nil
+		}
+
+		buf, err := hex.DecodeString(v)
+		if err != nil {
+			return fmt.Errorf("failed to convert string to byte: %w", err)
+		}
+
+		f.value = buf
+	case *string:
+		if v == nil {
+			f.value = nil
+			return nil
+		}
+
+		buf, err := hex.DecodeString(*v)
+		if err != nil {
+			return fmt.Errorf("failed to convert string to byte: %w", err)
+		}
+
+		f.value = buf
+	case []byte:
+		f.SetBytes(v)
+	case *[]byte:
+		if v == nil {
+			f.value = nil
+			return nil
+		}
+		f.SetBytes(*v)
+	default:
+		return fmt.Errorf("data does not match required *Binary or (string, *string, []byte, *[]byte) type")
```

Comment:
@mfdeveloper508 I suggest us to rework the method a little bit. First, we always set f.value to nil if `v` has zero value and return. So we can do this only once and the rest of the code will not have `if` statements. It seems the following code should work:


```go
func (f *Binary) Marshal(v interface{}) error {
	if reflect.ValueOf(v).IsZero() {
		f.value = nil
		return nil
	}

	switch val := v.(type) {
	case *Binary:
		f.value = val.value
	case string:
		buf, err := hex.DecodeString(val)
		if err != nil {
			return utils.NewSafeError(fmt.Errorf("converting string (as hex) to byte: %w", err), "failed to convert string (as hex) to byte") 
		}
		f.value = buf
	case *string:
		buf, err := hex.DecodeString(*val)
		if err != nil {
			return utils.NewSafeError(fmt.Errorf("converting string (as hex) to byte: %w", err), "failed to convert string (as hex) to byte") 
		}
		f.value = buf
	case []byte:
		f.SetBytes(val)
	case *[]byte:
		f.SetBytes(*val)
	default:
		return fmt.Errorf("data does not match required *Binary or (string, *string, []byte, *[]byte) type")
	}

	return nil
}
```